compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -702,19 +707,30 @@ class Semantic {
    *
    *  This method only handles cache logic and delegates the work to `cases`.
    */
-  def eval(expr: Tree, thisV: Addr, klass: ClassSymbol, cacheResult: Boolean = false): Contextual[Result] = log("evaluating " + expr.show + ", this = " + thisV.show, printer, res => res.asInstanceOf[Result].show) {
+
+  type evalCache = EqHashMap[Tree, Result]
+  def inputCache(using caches: (evalCache, evalCache)) = caches._1
+  def outputCache(using caches: (evalCache, evalCache)) = caches._2
```

Comment:
Type definition needs to be capitalized. I think opaque types will make the API nicer.

BTW, let's move domain definitions to the domain section.

Reply:
The cache key is incorrect: from the paper, it says `Expr -> Value -> Value` instead of `Expr -> Result`.